### PR TITLE
std-bit-slice implementation

### PR DESCRIPTION
**Pull request by anthonyabeo**
This PR addresses the implementation of the proposed bit-slicing operation per https://github.com/calyxir/calyx/issues/497 .

### CODE REVIEW

**Comment by rachitnigam** [comment]
Okay, there are failures in the interpreter and the compiler. @anthonyabeo can you take a shot at fixing the compiler errors? With `runt`, you can run a specific test using:
```
runt -i <test path> -d
```

If the output looks correct, you can save the updated file using the `-s` flag. You can [read more in the runt documentation](https://docs.rs/runt/latest/runt/). The failure in the interpreter comes from the missing implementation of the new primitive. By default, the interpreter tries to run all of the tests that the compiler runs. You can disable this by adding a `.skip` file in the location corresponding to the interpreter's expected output location.

**Comment by EclecticGriffin** [comment]
Zulip message about how to implement a new primitive in Cider: https://calyx.zulipchat.com/#narrow/stream/423600-development/topic/please-review/near/420297504

**Comment by anthonyabeo** [comment]
@EclecticGriffin thanks for the assist. I was able to resolve the interpreter's issue.

@rachitnigam noted. 

My main issue appears to be creating the local environment so I can get the tests to pass. On Mac, even the core tests are failing. I tried using Ubuntu 22.04 and cannot run `fud` (see the screenshot)

Screenshot from 2024-02-08 12-58-53

**Comment by rachitnigam** [comment]
You need to install the `calyx-py` module: https://docs.calyxir.org/builder/calyx-py.html

**Comment by anthonyabeo** [comment]
@rachitnigam @EclecticGriffin I have been able to install most of the dependencies. see screenshot
Screenshot from 2024-02-12 09-15-07

However, some of the tests are still failing and I am not sure how to go about debugging them. See the screenshot below for the output of `runt -i tests/frontend/dahlia/memory.fuse -d`
Screenshot from 2024-02-12 09-28-15

**Comment by anthonyabeo** [comment]
@rachitnigam I managed to resolve the failing tests. Not sure why but I still get failing tests locally when I merged `main` in this branch.

File: tests/correctness/std-bit-slice.futil [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+component main(@go go: 1) -> (@done done: 1) {
+  cells {
+    @external mem = comb_mem_d1(32, 1, 1);
+    slice = std_bit_slice(32, 10, 14, 4);
+    pad = std_pad(4, 32);
+  }
+
+  wires {
+    group bit_slice {
+      slice.in = 32'd206256;
```
Comment (rachitnigam) [inline tests/correctness/std-bit-slice.futil:13]: Can we write this number in binary so its more obvious what the slicing is doing?

Comment (rachitnigam) [reply tests/correctness/std-bit-slice.futil:13]: The syntax in Calyx is `32'b0101011` (note the `b` instead of `d`)

**Comment by rachitnigam** [comment]
Ah, I can really help debug those without the outputs but this PR seems to be passing all tests! There is one comment about updating the test but after that we can merge!

File: tests/correctness/std-bit-slice.futil (hunk 1 above)
Comment (anthonyabeo) [reply tests/correctness/std-bit-slice.futil:13]: sure

File: tests/correctness/std-bit-slice.futil.data [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "mem": {
+    "data": [10],
```
[the remark this answers is not in the document]
Comment (anthonyabeo) [reply tests/correctness/std-bit-slice.futil.data:3]: sure.

**Comment by anthonyabeo** [comment]
> Ah, I can really help debug those without the outputs but this PR seems to be passing all tests! There is one comment about updating the test but after that we can merge!

sure. Will do

**Pull request merged**